lib/src/visitor/serialize.dart

```diff
@@ -1287,44 +1289,88 @@ class _SerializeVisitor
-      _buffer.writeCharCode($rbrace);
-      return;
-    }
 
-    _writeLineFeed();
-    CssNode? previous_;
-    _indent(() {
-      for (var i = 0; i < children.length; i++) {
-        var child = children[i];
-        if (_isInvisible(child)) continue;
+    CssNode? prePrevious_;
+    CssNode previous = parent;
+    for (var child in children) {
+      if (_isInvisible(child)) continue;
+      if (previous != parent && _requiresSemicolon(previous)) {
+        _buffer.writeCharCode($semicolon);
+      }
 
-        var previous = previous_; // dart-lang/sdk#45348
-        if (previous != null) {
-          if (_requiresSemicolon(previous)) _buffer.writeCharCode($semicolon);
-          _writeLineFeed();
-          if (previous.isGroupEnd) _writeLineFeed();
-        }
-        previous_ = child;
+      if (_isTrailingComment(child, previous)) {
+        _writeOptionalSpace();
+        _withoutIndendation(() => child.accept(this));
+      } else {
+        _writeLineFeed();
+        _indent(() {
+          child.accept(this);
+        });
+      }
 
-        child.accept(this);
+      prePrevious_ = previous;
+      previous = child;
+    }
+
+    if (previous != parent) {
+      if (_requiresSemicolon(previous) && !_isCompressed) {
+        _buffer.writeCharCode($semicolon);
       }
-    });
 
-    if (_requiresSemicolon(previous_!) && !_isCompressed) {
-      _buffer.writeCharCode($semicolon);
+      if (prePrevious_ == parent &&
+          _isTrailingComment(previous, prePrevious_!)) {
+        _writeOptionalSpace();
+      } else {
+        _writeLineFeed();
+        _writeIndentation();
+      }
     }
-    _writeLineFeed();
-    _writeIndentation();
+
     _buffer.writeCharCode($rbrace);
   }
 
   /// Whether [node] requires a semicolon to be written after it.
   bool _requiresSemicolon(CssNode? node) =>
       node is CssParentNode ? node.isChildless : node is! CssComment;
 
+  /// Whether [node] represents a trailing comment when it appears after
+  /// [previous] in a sequence of nodes being serialized.
+  ///
+  /// Note [previous] could be either a sibling of [node] or the parent of
+  /// [node], with [node] being the first visible child.
+  bool _isTrailingComment(CssNode node, CssNode previous) {
+    // Short-circuit in compressed mode to avoid expensive span shenanigans
+    // (shespanigans?), since we're compressing all whitespace anyway.
+    if (_isCompressed) return false;
+    if (node is! CssComment) return false;
+
+    var previousSpan = previous.span;
+    if (previous is CssParentNode && previous.children.contains(node)) {
+      // Walk back from just before the current node starts looking for the
+      // parent's left brace (to open the child block).  This is safer than
+      // a simple forward search of the previousSpan.text as that might contain
+      // other left braces.
+      var searchFrom = node.span.start.offset - previousSpan.start.offset - 1;
+      if (searchFrom < 0) {
```

Comment:
This isn't necessarily the most reliable determiner of whether the math on the span offsets is nonsense... it's possible that `node.span` appears late in a different file than `previousSpan`, for example. It's probably better to write a little `SpanExtensions.contains()` utility and check that `previous.span.contains(node.span)`. You could even use that instead of the `previous.children.contains(node)` check.

Reply:
WIP

Reply:
IIUC then I think this is done :)